L37-go-custom-transports.md

```diff
@@ -0,0 +1,415 @@
+package server
+
+// TransportListener provides Streams to be handled by grpc.
+type TransportListener interface {
+	// Accept blocks until a new Stream is created.  Returns an error when the
+	// listener should stop being used.
+	Accept() (Stream, error)
+
+	// GracefulClose causes the TransportListener to stop accepting new
+	// incoming streams, and returns when all outstanding streams have
+	// completed.  Pending and future calls to Accept should return a non-nil
+	// error.
+	GracefulClose()
+
+	// Close immediately closes all outstanding connections and streams.
+	// Pending and future calls to Accept should return a non-nil error.
+	Close()
+}
+
+// Header defines the fields set in the client header.
+type ClientHeader struct {
+	MethodName string
+	Metadata   metadata.MD
+}
+
+// Trailer defines the information contained in the server's trailer.
+type Trailer struct {
+	Status   status.Status
+	Metadata metadata.MD
+}
+
+// Stream defines all necessary RPC functions.
+type Stream interface {
+	// RecvHeader returns the client's header.
+	RecvHeader() ClientHeader
+
+	// RecvMsg receives the next message on the Stream into m and returns true
+	// unless the Stream terminates before a full message is received.
+	RecvMsg(m transport.IncomingMessage) bool
+	// SendHeader immediately sends the header to the client.  Returns true
+    // unless the Stream has encountered an error.
+	SendHeader(Header) bool
+	// SendMsg queues the message m to be sent by the Stream and returns true
+	// unless the Stream terminates before queuing the message.
+	SendMsg(m transport.OutgoingMessage, opts StreamSendMsgOptions) bool
+	// Close terminates the RPC and sends the trailer provided.  All subsequent
+	// operations on the Stream except Info should fail.
+	Close(Trailer)
+
+	// Info returns information about the stream's current state.
+	Info() StreamInfo
+}
+
+// Header contains header data sent by the server.
+type Header struct {
+	Metadata metadata.MD
+}
+
+// StreamInfo contains information about the stream's current state.  All
+// information is optional.
+type StreamInfo struct {
+	// RemoteAddr is the address of the client (typically an IP/port).
+	RemoteAddr string
+}
+```
+
+
+## Rationale
+
+Alternatives considered:
+
+- Hardcode other transports, like the in-memory transport.
+
+  This does not allow users the flexibility of implementing their own transports
+  and requires grpc-go to maintain many separate transports.
+
+- Require byte-based transports.
+
+  This does not allow for a more efficient in-memory transport that is capable
+  of copying proto messages rather than serializing and deserializing.
+
+- Inject the client `TransportBuilder` directly into the `ClientConn`, rather
+  than allowing the resolver and balancer to influence it.
+
+  This would have a simpler API, however, it would require each `ClientConn` to
+  use only a single transport.  The proposed design enables a `ClientConn` with
+  different types of connections.
```

Comment:
The value of this is unclear. Under what cases would a single connection want different types of connection?

The only thing I can imagine might be to provide both in-process and remote connections for a service that is exposed both locally and by the server's peers. But that would only be useful if there were a rich API for affinity -- such that some requests (based on affinity keys) might be routed locally over an in-process transport and others could be routed to a remote server. But that could also easily be implemented as a "composite channel" abstraction -- where a channel delegates to a `ClientConn` with in-process connections for some requests but to another `ClientConn` for others.

Perhaps a simpler API would be more valuable?